### ISSUE TITLE
API GET /documents doesn't return all information

### ISSUE THREAD

**Issue by battosai30** [Dolibarr/dolibarr#29439]
### Bug

Hi,

I use API to retrive share link of my invoices. After some tests it seemed OK but I don't know why, GET /documents returned content is not consistent.

Exampel using API explorer : 

GET api/index.php/documents?modulepart=invoice&id=226

```
[
  {
    "name": "FA2403-0183.pdf",
    "path": "/var/lib/dolibarr/documents/facture/FA2403-0183",
    "level1name": "FA2403-0183",
    "relativename": "FA2403-0183.pdf",
    "fullname": "/var/lib/dolibarr/documents/facture/FA2403-0183/FA2403-0183.pdf",
    "date": 1710515325,
    "size": 67051,
    "type": "file",
    "label": "XXXXXXXXXX",
    "entity": "1",
    "filename": "FA2403-0183.pdf",
    "filepath": "facture/FA2403-0183",
    "fullpath_orig": "",
    "description": "",
    "keywords": "",
    "cover": null,
    "position": "1",
    "gen_or_uploaded": "generated",
    "extraparams": null,
    "date_c": 1710515325,
    "date_m": 1710515325,
    "fk_user_c": "1",
    "fk_user_m": null,
    "acl": null,
    "src_object_type": "facture",
    "src_object_id": "226",
    "id": "682",
    "ref": "682",
    "share": "XXXXXXXXXXXX"
  }
]
```

GET api/index.php/documents?modulepart=invoice&id=251
```

[
  {
    "name": "FA2404-0204.pdf",
    "path": "/var/lib/dolibarr/documents/facture/FA2404-0204",
    "level1name": "FA2404-0204",
    "relativename": "FA2404-0204.pdf",
    "fullname": "/var/lib/dolibarr/documents/facture/FA2404-0204/FA2404-0204.pdf",
    "date": 1713257786,
    "size": 67550,
    "type": "file"
  }
]
```

Both in webUI have a share link for example.

I can't get what could lead to this difference of behaviour ...

Regards

### Dolibarr Version

18.0.4

### Environment PHP

7.4

### Environment Database

MariaDB

## EDIT

Ok I'm close from a solution :)

I found the problem but don't know exactly how to solve it ... And maybe there are two bugs in one

So in /htdocs/api/class/api_documents.class.php there is the main function, which call for fetchAll() of htdocs/ecm/class/ecmfiles.class.php 
The problem here is that the SQL built use "LIKE" in WHERE condition. In my case, I'm looking for an invoice. But if the src_object_id exists for different objects, multiple lines are return. So maybe it's the bug n°1 because "LIKE %facture%" also returns "facture_fourn" type for example. 

So if multiple lines are returned by fetchAll() , this code is invloved : 

```
$result = $ecmfile->fetchAll('', '', 0, 0, array('t.src_object_type' => $objectType, 't.src_object_id' => $object->id));
				if ($result < 0) {
					throw new RestException(503, 'Error when retrieve ecm list : '.$this->db->lasterror());
				} elseif (is_array($ecmfile->lines) && count($ecmfile->lines) > 0) {
					$count = count($filearray);
					for ($i = 0 ; $i < $count ; $i++) {
						if ($filearray[$i]['name'] == $ecmfile->lines[$i]->filename) {
							$filearray[$i] = array_merge($filearray[$i], (array) $ecmfile->lines[0]);
						}
					}
				}
```

And it seems not able to healthy manage this case because I debuged it, if ($filearray[$i]['name'] == $ecmfile->lines[$i]->filename) is never triggered (I'm still searching for a solution on this part).

So it's bug n°2 

## EDIT 2

Solved :)

```
if ($filearray[$i]['name'] == $ecmfile->lines[$i]->filename) {
							$filearray[$i] = array_merge($filearray[$i], (array) $ecmfile->lines[0]);
}

```
Replaced by 
```
foreach ($ecmfile->lines as $line) {
                                                        if ($filearray[$i]['name'] == $line->filename) {
                                                                $filearray[$i] = array_merge($filearray[$i], (array) $line);
                                                }
}
```
Allows to scan all files returned in $ecmfiles for each file in $filearray, and merge them if filenames match

**Comment by JonBendtsen** [comment]
@battosai30 

did you fix it in your version 18.0.4 on in the develop that you can see here
https://github.com/Dolibarr/dolibarr/blob/develop/htdocs/api/class/api_documents.class.php

and here
https://github.com/Dolibarr/dolibarr/blob/develop/htdocs/ecm/class/ecmfiles.class.php

**Comment by battosai30** [comment]
@JonBendtsen no I did not I just read the contributing rules in order to do things well (it's my first real contribution) ;)
